### PR TITLE
Fix #2284 must be sent from authorization page

### LINKED ISSUE #2284: Authorization form must be sent from authorization page
Hi! after upgrading to master branch (around 40 days of new commits) admins are no longer able to impersonate users, we get this error right after clicking "Authorize" on the new page:

https://github.com/jupyterhub/jupyterhub/blob/5b4f0d4304b1ea57b4d398609f2777bd93131ef4/jupyterhub/apihandlers/auth.py#L255

```
403 : Forbidden
Authorization form must be sent from authorization page
```

We are still using GitHubOAuthenticator and Postgres as backend.

Also other strange thing is happening, the admin page is not properly listing all users, but this is probably related to some of our custom page.html templates:

image

### PR DESCRIPTION
**Pull request by elgalu**
Fix #2284 Authorization form must be sent from authorization page

### CODE REVIEW

**Comment by manics** [comment]
Is it possible for this to open up a security hole, e.g. by allowing an attacker to force a user to use http instead of https?

**Comment by elgalu** [comment]
Hi, I opened this PR to help with https://github.com/jupyterhub/jupyterhub/issues/2284#issuecomment-713670155 however I'm not a security expert to answer your question.

**Comment by manics** [comment]
I opened an experimental PR at https://github.com/jupyterhub/configurable-http-proxy/pull/267. That needs to be reviewed for security implications, but might also work? Are you able to give it a go? Most (or all) of the current maintainers don't have a platform where we can reproduce the error.

**Comment by chicocvenancio** [comment]
> Is it possible for this to open up a security hole, e.g. by allowing an attacker to force a user to use http instead of https?

I don't think so (?). 

@minrk should take a look at this, but I am not opposed to @elgalu's approach here. 

Tornado is checking origin before it gets to this, and the endpoint is authenticated, I do not see any holes added by stripping the protocol here.
In particular, the deployment should be enforcing https in front of jupyterhub and this really would not be impacted.


#2284 is about JupyterHub seeing the request as http while the browser is using https. In theory we could allow only that combination, but I do think it will make the code significantly less readable and will impact security either way.

**Comment by elgalu** [comment]
@manics even if your fix at configurable-http-proxy works, I think most enterprise users do not use configurable-http-proxy so that alone won't solve this issue (for everyone)

**Comment by minrk** [comment]
This *does* open a security issue, in that what appears to be a cross-origin request (two websites that share a domain, one of which could be compromised, especially since it is by definition insecure) would no longer be blocked.

The problem in #2284 is in failure to *identify* that an HTTPS request is actually HTTPS because of the multiple proxied hops, and some layers not correctly implementing proxy-protocols (I get lost in terms of who is actually doing something wrong!).

I think we can combine the approach here with what is in https://github.com/jupyterhub/configurable-http-proxy/pull/267 and allow the http request *if and only if* https is found in the proxy headers (e.g. x-forwarded-proto, Forwarded).

However, because we do things like set cookies with `secure=true`, if we got this far, then we actually do know that the browser (if it implements cookies correctly and defaults are being used) sent the request over https.

While it's not ideal, getting proxies to agree on what's being forwarded and how to determine the actual URL being requested has turned out to be between hard and impossible. So I think with the other measures we have in place, only considering the host is perhaps enough here.

**Comment by minrk** [comment]
I also think that we can benefit from the fact that when this fails and things are working right, it's because the request is http and the origin is https. (proxies messed up request url, but Referer was left alone and is set by the browser, which is whose perspective we care about)

File: jupyterhub/apihandlers/auth.py
Comment (minrk) [inline jupyterhub/apihandlers/auth.py:278]:
```suggestion
        # trim protocol, which cannot be trusted with multiple layers of proxies anyway
        # Referer is set by browser, but full_url can be modified by proxy layers to appear as http
        # when it is actually https
        referer_proto, _, stripped_referer = referer.partition("://")[2]
        referer_proto = referer_proto.lower()
        req_proto, _, stripped_full_url = full_url.partition("://")[2]
        req_proto = req_proto.lower()
        if referer_proto != req_proto:
            self.log.warning("Protocol mismatch: %s != %s", referer, full_url)
            if req_proto == "https":
                # insecure origin to secure target is not allowed
                raise web.HTTPError(403, "Not allowing authorization form submitted from insecure page")
```

**Review by minrk — COMMENTED** [review]
@elgalu I added a suggestion to restrict the error to http->https requests and allow https->http, which, if I understand correctly, is ~always how these proxy issues appear. Can you verify that this will work for you?

**Review by minrk — APPROVED** [review]

**Comment by minrk** [comment]
@elgalu if you can verify that this change still works for you, I'll go ahead and merge it.

**Comment by elgalu** [comment]
Hi, I'll verify this tonight.

**Comment by elgalu** [comment]
Confirmed! this works in our custom environment that uses https://github.com/zalando-incubator/kubernetes-on-aws and https://github.com/zalando/skipper proxy. Thanks!!!

**Comment by minrk** [comment]
@elgalu Thank you for your testing and patience! :)

**Comment by elgalu** [comment]
my pleasure!!

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/set-up-multiple-jupyterhub-on-same-domain-with-different-ports-using-nginx/8706/6

**Pull request merged**
